logstash-core/lib/logstash/agent.rb

```diff
@@ -75,9 +75,12 @@ def execute
       Stud.interval(@reload_interval) { reload_state! }
     else
       while !Stud.stop?
-        if clean_state? || running_pipelines?
-          sleep 0.5
-        else
+        if running_user_defined_pipelines?
+          sleep(0.5)
+        elsif running_pipelines?
+          logger.debug("Shutting down system pipelines")
+          shutdown_pipelines
+        elsif clean_state? || !running_pipelines?
```

Comment:
I think `clean_state?` must lead to sleep instead of break like in the previous code. This is because if a sighup reloads a pipeline and the pipeline aborts during a plugin.register, then logstash will terminate.